src/bhfmm/HaloBufferOverlap.h

```diff
@@ -92,15 +93,13 @@ _areaBuffers(areaNumber), _edgeBuffers(edgeNumber), _cornerBuffers(cornerNumber)
 	}
 
 	if(areaNumber != 0){
-
-		_areaRequests = new MPI_Request[_areaBuffers.size()];
-//		std::cout << "areaBufferSize: "<<_areaBuffers.size() << "\n";
+		std::vector<MPI_Request> _areaRequests(_areaBuffers.size());
```

Comment:
this creates a new variable `_areaRequests` that shadows the existing class member. However, this new variable only exists inside this `if`, rendering it useless.

The constructor in this class is written rather horribly. I _guess_ an OK-ish way to do achieve what you want is to:
1. change the data type of the member at its declaration
2. remove these `if`s
3. Initialize the member in the member initialization list of the constructor like this: ` _areaRequests(areaNumber)`. (should have the same effect as until now because _areaBuffers are initialized with a size of `areaNumber`.)
4. Whenever this vector is passed to a (MPI) function that expects a pointer you can pass `_areaRequests.data()` to get the pointer of the actual data field.

Reply:
I implemented this in the latest commit.
During the implementation, I also saw [this line](https://github.com/ls1mardyn/ls1-mardyn/blob/339606af0d9a91431ed3a1c1b1642618fb2483a1/src/bhfmm/HaloBufferOverlap.h#L195). It seems like this would be an alternative and the original way of deleting `_areaRequests`, etc. I have no idea why it is commented out 😆

Reply:
Disclaimer: I'm not an expert on this.

As far as I understand, we should call `MPI_Request_free()` on all elements of those vectors *either way*. This does not deallocate the array (this is done by its destructor) but the request handles in it (afaik we can think of them as pointers?). So the master should have a double memory leak here:
1) The c-style array of _areaBuffers that is not freed (as far as I can see).
2) The memory for the requests itself.

Reply:
I tried to run a simulation with the commented-in `MPI_Request_free()` commands. However, this led to the following error:
```bash
*** The MPI_Request_free() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
```
So maybe the free-ing of the `MPI_Request` is handled by the `MPI_finalize` command? Should we just `default` the destructor (we have smart pointers now, so point 1 of your reply should be handled correctly now)?

Side note:
It is not very trivial (of course!) to use `bhfmm`. After looking up the xml syntax for it in the header of [this file](https://github.com/ls1mardyn/ls1-mardyn/blob/27625f5d120326447ede42fe322dcda93f1dbc39/src/bhfmm/FastMultipoleMethod.h#L57-L60), I tried to _replace_ the `reaction field` electrostatic section in a config with the FastMultipoleMethod electrostatic section. Though, this doesn't work (`ERROR: Electrostatics section for reaction field setup missing.`) and the reaction field section is needed additionally. Inspect [this code](https://github.com/ls1mardyn/ls1-mardyn/blob/27625f5d120326447ede42fe322dcda93f1dbc39/src/Simulation.cpp#L283-L306) to see why. I am not really deep into the bhfmm method, but this does not look correct (or at least intuitive) to me.

Reply:
After moving all object deletions from the destructor of `Simulation` to the `finalize` method, I encountered the following error:
`MPI_ERR_REQUEST: invalid request`
It seems like the requests were already freed before. This could also be tested with something like:
`MPI_Test(&request, &flag, &status);`

Something like this could work, but is highly inelegant in my opinion:
```cpp
// Test if the request has already been completed
int flag;
MPI_Status status;
MPI_Test(&_areaRequests[i], &flag, &status);
if (flag) {
    std::cout << "Request already completed, no need to free." << std::endl;
} else {
    // Free the request if it is still active
    MPI_Request_free(&_areaRequests[i]);
    _areaRequests[i] = MPI_REQUEST_NULL;  // Avoid double-freeing
}
```

Reply:
Have a look at `MarDyn.cpp::main()`. It is roughly structured like this:

```
int main(int argc, char** argv) {
	MPI_Init(&argc, &argv);
	// ...
	Simulation simulation;
	// ...
	MPI_Finalize();
}
```

This means that Simulation is initialized after the `MPI_Init()`, but its destructor AFTER `MPI_Finalize()` because this is called when `simulation` goes out of scope, which is at the closing bracket at the end of the `main()` function. A super easy and somewhat elegant fix would be to introduce a block from `MPI_Init()` to `MPI_Finalize()` so everything is destructed before we call `MPI_Finalize()`:

```
int main(int argc, char** argv) {
	MPI_Init(&argc, &argv);
	{
		// ...
		Simulation simulation;
		// ...
	}
	MPI_Finalize();
}
```

Reply:
It seems I've misunderstood `MPI_Request_free()`. Apparently we only have to call this if the requests have never been used, or if they are persistent requests that are not cleaned up by the communication functions. I've no idea how they are used here but the behavior you observed does not suggest that either of these is the case.
Thus I guess it's fine if we just don't do any `MPI_Request_free()`?

Reply:
Yes, those are the 2 options I see:
1. scoping as you suggested (drawback: closing bracket ~100 lines below opening one could be confusing)
2. Explicit clean up function/deletions during `finalize` (drawback: could be  errorprone)
We could go for 1.

Regarding `MPI_Request_free()`:
Sounds good and I this it works to not explicitly call it.

Aren‘t the elements of the arrays (eg _`areaBuffer`) deleted in the destructor?

Reply:
> Aren't the elements of the arrays (eg _areaBuffer) deleted in the destructor?

Now that you have changed them to `std::vector,` yes. The old C-style arrays are just pointers, and whose memory was allocated manually via `new` not.

Reply:
I have changed the `_areaRequests` not the `_areaBuffers`.
The _areaBuffers are:
```cpp
//arrays for MPI halo transfer (send)
std::vector<T *>  _areaBuffers,  _edgeBuffers,  _cornerBuffers;
```

Reply:
Ah, sorry, my bad.

Usually, `delete[]` has to be used for data that was allocated with `new[]`, which seems to be the case for each element of `_areaBuffers`. So yes, this is still needed.

However, I'd actually why we have vectors of pointers instead of just `std::vector<std::vector<T>>`. Then we could just make use of the regular vector destructors and not care about manual memory management.

Reply:
I also changed the `_areaBuffers` et al. now. This required some further changes in a few other files.